pkg/torcx/propagate.go

```diff
@@ -15,94 +15,197 @@
-	srcDir := filepath.Join(imageRoot, "usr", "lib", "systemd")
-	st, err := os.Stat(srcDir)
-	if os.IsNotExist(err) {
-		return nil
-	} else if err != nil {
-		return err
+func retrieveAssets(applyCfg *ApplyConfig, imageRoot string) (*Assets, error) {
+	if applyCfg == nil {
+		return nil, errors.New("missing apply configuration")
 	}
-	if !st.IsDir() {
-		return nil
+	if imageRoot == "" {
+		return nil, errors.New("missing image top directory")
+	}
+	assets := &Assets{}
+	path := filepath.Join(imageRoot, manifestPath)
+	_, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// Corner-case: missing manifest, no assets to propagate
+			return assets, nil
+		}
+		return nil, err
 	}
 
-	_, err = os.Stat(SYSTEMD_DIR)
+	var manifest ImageManifestV0
+	b, err := ioutil.ReadFile(path)
 	if err != nil {
-		return errors.Wrapf(err, "error checking for systemd volatile directory %s", SYSTEMD_DIR)
+		return nil, err
+	}
+	if err := json.Unmarshal(b, &manifest); err != nil {
+		return nil, err
+	}
+
+	return &manifest.Value, nil
+}
+
+// propagateServiceUnits installs systemd service unit and unit-like files as runtime
+// units in /run/systemd/system/.
+func propagateServiceUnits(applyCfg *ApplyConfig, imageRoot string, services []string) error {
+	if len(services) <= 0 {
+		// Corner-case: no services to propagate
+		return nil
+	}
+	if applyCfg == nil {
+		return errors.New("missing apply configuration")
+	}
+	if imageRoot == "" {
+		return errors.New("missing image top directory")
 	}
 
-	// Walk the source directory
-	err = filepath.Walk(srcDir, func(srcPath string, srcStat os.FileInfo, err error) error {
-		if err != nil {
+	servicesDir := filepath.Join(systemdDir, "system")
+	if _, err := os.Stat(servicesDir); err != nil {
+		return errors.Wrapf(err, "error checking for systemd runtime directory %s", servicesDir)
+	}
+	for _, servEntry := range services {
+		if servEntry == "" {
+			continue
+		}
+		path := filepath.Join(imageRoot, servEntry)
+		if err := flattenUnitAsset(applyCfg, servicesDir, path); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// propagateBins symlinks binaries from unpacked image to torcx bindir.
+func propagateBins(applyCfg *ApplyConfig, imageRoot string, binaries []string) error {
+	if len(binaries) <= 0 {
+		// Corner-case: no binaries to propagate
+		return nil
+	}
+	if applyCfg == nil {
+		return errors.New("missing apply configuration")
+	}
+	if imageRoot == "" {
+		return errors.New("missing image top directory")
+	}
+
+	for _, binEntry := range binaries {
+		if binEntry == "" {
+			continue
+		}
+		path := filepath.Join(imageRoot, binEntry)
+		if err := flattenBinAsset(applyCfg, applyCfg.RunBinDir(), path); err != nil {
 			return err
 		}
+	}
+	return nil
+}
+
+// flattenBinAssets propagates a single binary or a directory of binaries,
+// flattening all intermediate directories.
+func flattenBinAsset(applyCfg *ApplyConfig, binDir string, asset string) error {
```

Comment:
Bikeshedding: The name "flatten" is a bit confusing to me. I associate it with collection code, and don't expect it to have side-effects. Maybe call it `propagateBinAsset`?